tests/test_ingest.py

```diff
@@ -0,0 +1,154 @@
+
+EXAMPLES
+
+# Run all tests
+pytest
+
+# Run tests with names containing the string "dense"
+pytest -k "dense"
+
+# Run all tests in a manner that shows any print() statements
+python3 test_ingest.py
+
+# Run all tests, using multiple CPUs
+# Details: https://pypi.org/project/pytest-xdist/
+pytest -n auto
+
+# Run all tests, show code coverage metrics
+# For explanation of coverage report, see:
+# https://coverage.readthedocs.io/en/v4.5.x/branch.html
+coverage run --branch test_ingest.py; coverage report -m --include *scp-ingest-pipeline/ingest*
+
+"""
+
+import ast
+from glob import glob
+import sys
+import unittest
+from unittest.mock import patch
+
+import google.cloud
+
+sys.path.append('../ingest')
+from ingest_pipeline import *
+
+def mock_load_expression_data(self, *args, **kwargs):
+    """Enables overwriting normal function with this placeholder.
+    Returning the arguments enables tests to verify that the code invokes
+    this method with expected argument values.
+
+    Ideally we would mock Firestore itself, but existing libraries to mock
+    Firestore (e.g. https://github.com/mdowds/python-mock-firestore) don't
+    implement important Firestore methods like "batch()".
+    """
+    self.load_expression_data_args = args
+    self.load_expression_data_kwargs = kwargs
+
+def mock_firestore_client(*args, **kwargs):
+    """Mocks firestore.Client() by returning nothing upon initializing client
+    """
+    return
+
+def get_nth_gene_models(n, models, mock_dir):
+    """Return Nth actual and expected gene models, using actual and mock data
+    """
+    # TODO: Dense loads models as a `list`, Mtx loads models as a `dict_values`
+    # It seems both would ideally load using the same type.  Reconcile.
+    if isinstance(models, list):
+        # For Dense
+        actual_model = models[n].__dict__
+    else:
+        # For Mtx
+        actual_model = list(models)[n].__dict__
+
+    # Uncomment to print out new baseline data
+    # Process to update baselines is manual: copy and paste it into new file
+    # TODO: Automate when reasonable
+    # print('actual_model')
+    # print(actual_model)
+
+    with open(f'mock_data/{mock_dir}/gene_model_{n}.txt') as f:
+        # Create a dictionary from the string-literal mock
+        expected_model = ast.literal_eval(f.read())
+    
+    return actual_model, expected_model
+
+# Mock method that loads data to Firestore
+IngestPipeline.load_expression_data = mock_load_expression_data
+
+
+class IngestTestCase(unittest.TestCase):
+
+    def setup_ingest(self, args):
+        args_list = args.split(' ')
+        parsed_args = create_parser().parse_args(args_list)
+        validate_arguments(parsed_args)
+        arguments = vars(parsed_args)
+
+        ingest = IngestPipeline(**arguments)
+
+        if hasattr(ingest, 'ingest_expression'):
+            getattr(ingest, 'ingest_expression')()
+
+        return ingest
+
+    @patch('google.cloud.firestore.Client', side_effect=mock_firestore_client)
```

Comment:
So here you would have 3 cases (give or take a test or 2) for dense matrices. One for ingesting dense matrices correctly (test_ingest_expression_correct, test_ingest_expression_raises_InvalidArgument_correcty) and incorrectly (test_ ingest_expression_raises_exceptions_incorrectly). You would Mock the files and set the return file to a fiven value. I can explain this in further detail.

Reply:
Great suggestion on testing multiple cases.  That will increase coverage, soundly.  Let's touch base; I plan to push this refinement today.

Reply:
155b088 increases coverage per this comment.